### ISSUE TITLE
overriding level_to_s

### ISSUE THREAD

**Issue by gingerlime** [rocketjob/semantic_logger#56]
I have a quick question about the "official" way to get the log level to show fully (e.g. `ERROR` instead of `E`). I can see that this is done in `level_to_s` method, and used in the default formatter.

Would it be best to monkey-path this method? (how?) or to define a custom formatter?

Intuitively I would use a custom formatter, but feels like a lot of copy&paste for changing a small fraction of the default format.

Or maybe there's a different option altogether?

**Comment by gingerlime** [comment]
(as a side note I'm curious about the design choice to shrink the level to one letter. I've never seen it done before, but I'm not familiar with too many logging frameworks)

**Comment by reidmorrison** [comment]
Yes, the way to customize the logging format: https://rocketjob.github.io/semantic_logger/customize.html

It would be nice to be able to "configure" just one aspect of the log format without having to supply the complete block. Want to submit a PR with that capability since I am sure you are not the only one that would like it?

The real indication of the log level is actually the color of the line itself, which  is much faster than trying to read the log level text of every entry.

You can still grep for specific levels as follows
~~~
tail -f log/production.log | egrep "\[1m(W|E|F).\[0m”
~~~

**Comment by gingerlime** [comment]
Thanks @reidmorrison. I'd be happy to submit another PR, but not entirely sure how to make this "pluggable"... I'll think about it.

Yeah, colors are great, and I can regex this thing, but it's mostly for matching things automatically on logentries and generating alerts, so I'd rather rely on something a bit more expressive (plus, we remove colors on production, so logs are cleaner to inject).

**Comment by reidmorrison** [comment]
In our production environment we created colored logs for humans to read. We also write to a JSON formatted log file that is read by splunk.
As for alerting, Semantic Logger does that directly by feeding New Relic, Bugsnag and others in real time without the need to "inject" the logs in any way. https://rocketjob.github.io/semantic_logger/appenders.html

**Comment by gingerlime** [comment]
Yeah, it makes sense to do things this way. We do things a bit differently. We only read logs on logentries typically, so colors just mess up things there. We also want to write to log files rather than directly using the logentries appender for a few of reasons (Not all logs are written by Rails, and we anyway send other log files using the logentries daemon, and it's less vendor-locked).

We use sentry for alerts, and the sentry appender works great for this. But some alerts are better done with logentries. For example, velocity checks. We don't log every warning to sentry, but we do write it to the log. We have a rule that detects a high number of warnings and alerts us. This alert is set on logentries.

In any case, just to give some context. I can certainly change the regex on logentries to capture the condensed log-level, but I just thought having the full level is more common in logs, and doesn't really waste that much space or cognitive power to read through.

Semantic logger looks great, and the appenders and custom formatters gives us flexibility. Just wondered if there's a simpler hack or a better way to tweak this particular part of the log. I'll keep thinking if there's a simple way to structure this so it's easier to hook into, but I can't think of anything right now.

Thanks for taking the time to answer, @reidmorrison. Much appreciated.

**Comment by gingerlime** [comment]
I'll close this for now, because my question was answered. I'd be happy to continue this discussion however and contribute to the project.

**Comment by reidmorrison** [comment]
Recommend looking into Bugsnag, we send all of our errors to Bugsnag which has builtin velocity/spike detection. It is great at managing errors and exceptions across all of our servers for a very reasonable cost. Definitely helped us reduce errors and track new ones.

**Comment by gingerlime** [comment]
Thanks @reidmorrison. Bugsnag looks interesting. We recently switched to Sentry, so I don't think we're likely to switch again (Although I can't say I'm super happy with their recent pricing changes... )

**Comment by reidmorrison** [comment]
We refactored the formatters in master to allow for easier customization of the format.


Create a custom formatter:
~~~ruby
class MyFormatter < SemanticLogger::Formatters::Default
  # Return the complete log level name in uppercase
  def level
    log.level.upcase
  end
end
~~~

Specify the formatter when creating the appender:
~~~ruby
SemanticLogger.add_appender(io: $stdout, formatter: MyFormatter.new)
~~~

Example usage:
~~~ruby
Rails.logger.info "Hello World"

# => 2017-04-05 01:05:52.868286 INFO [13143:70216759638540 (irb):11] Rails -- Hello World
~~~

The formatter for an existing appender can also be changed directly if required.
~~~ruby
SemanticLogger.appenders.first.formatter = MyFormatter.new
~~~

**Comment by gingerlime** [comment]
Looks really cool! Thanks @reidmorrison 

Is this "officially" released?

**Comment by reidmorrison** [comment]
It will be released as part of the v4.1 gem, which will be created once this change has been tested and then deployed to our production environment to ensure there are no issues.
